src/kernel/scheduler.zig

```diff
@@ -0,0 +1,310 @@
+const Task = arch.Task;
+const Allocator = std.mem.Allocator;
+const TailQueue = std.TailQueue;
+
+/// The function type for the entry point.
+const EntryPointFn = fn () void;
+
+/// The default stack size of a task. Currently this is set to a page size.
+const STACK_SIZE: u32 = arch.MEMORY_BLOCK_SIZE / @sizeOf(usize);
+
+/// Pointer to the start of the main kernel stack
+extern var KERNEL_STACK_START: []u32;
+
+/// The current task running
+var current_task: *Task = undefined;
+
+/// Array list of all runnable tasks
+var tasks: TailQueue(*Task) = undefined;
+
+/// The memory allocator for creating the stack.
+var alloc: *Allocator = undefined;
+
+///
+/// The idle task that just halts the CPU but the CPU can still handle interrupts.
+///
+fn idle() noreturn {
+    arch.spinWait();
+}
+
+///
+/// Round robin. This will first save the the current tasks stack pointer, then will pick the next
+/// task to be run from the queue. It will add the current task to the end of the queue and pop the
+/// next task from the front as set this as the current task. Then will return the stack pointer
+/// of the next task to be loaded into the stack register to load the next task stack to pop off
+/// its state. Interrupts are assumed disabled.
+///
+/// Argument:
+///     IN ctx: *arch.CpuState - Pointer to the exception context containing the contents
+///                              of the registers at the time of a exception.
+///
+/// Return: usize
+///     The new stack pointer to the next stack of the next task.
+///
+pub fn pickNextTask(ctx: *arch.CpuState) usize {
+    // Save the stack pointer from old task
+    current_task.stack_pointer = @ptrToInt(ctx);
+
+    // Pick the next task
+    if (tasks.pop()) |new_task_node| {
+        // Get the next task
+        const next_task = new_task_node.data;
+
+        // Move some pointers to don't need to allocate memory, speeds things up
+        new_task_node.data = current_task;
+        new_task_node.prev = null;
+        new_task_node.next = null;
+
+        // Add the 'current_task' node to the end of the queue
+        tasks.prepend(new_task_node);
+
+        current_task = next_task;
+    } else {
+        // Not more tasks, this isn't possible as the idle task is noreturn
+        panic(@errorReturnTrace(), "No more tasks?\n", .{});
+    }
+
+    // Context switch in the interrupt stub handler which will pop the next task state off the
+    // stack
+    return current_task.stack_pointer;
+}
+
+///
+/// Create a new task and add it to the scheduling queue. No locking.
+///
+/// Arguments:
+///     IN entry_point: EntryPointFn - The entry point into the task. This must be a function.
+///
+/// Error: Allocator.Error
+///     OutOfMemory - If there isn't enough memory for the a task/stack. Any memory allocated will
+///                   be freed on return.
+///
+pub fn createNewTask(entry_point: EntryPointFn) Allocator.Error!void {
```

Comment:
Could this take a `Task` instead and add it to the queue? I think that would make for a cleaner and more flexible API. A name change could follow too (`queueTask`? `scheduleTask`?).

Reply:
Ah yes, I think there was a comment before, I will rework :)

Reply:
Nice, then it could be used to enqueue the old task when switching.

Reply:
I manipulate the queue pointers instead so to reduce the allocations and make the switching faster

Reply:
That's a nice optimisation. Forget the second bit then.